File: app/src/main/java/app/android/ttp/mikazuki/yoshinani/repository/retrofit/service/RetrofitUserService.java
package app.android.ttp.mikazuki.yoshinani.repository.retrofit.service;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.util.List;

import javax.annotation.Nonnull;

import app.android.ttp.mikazuki.yoshinani.repository.retrofit.entity.ResponseMessage;
import app.android.ttp.mikazuki.yoshinani.repository.retrofit.entity.User;
import retrofit2.Response;
import retrofit2.http.Body;
import retrofit2.http.DELETE;
import retrofit2.http.Field;
import retrofit2.http.FormUrlEncoded;
import retrofit2.http.GET;
import retrofit2.http.PATCH;
import retrofit2.http.POST;
import retrofit2.http.Path;
import retrofit2.http.Query;
import rx.Observable;

/**
 * @author haijimakazuki
 */
public interface RetrofitUserService {

    static final String PATH_USERS = "users";
    static final String PATH_USER = "users/{id}";
    static final String PATH_LOGIN = "users/sign_in";
    static final String PATH_USER_SEARCH = "users/search";
    static final String PATH_PASSWORDS = "passwords";
    static final String PATH_PASSWORDS_INIT = "passwords/init";
    static final String PATH_PASSWORDS_RESET = "passwords/reset";
    static final String PATH_NOTIFICATION_TOKENS = "notification_tokens";
    static final String PATH_OAUTH = "oauth_registrations";

    @POST(PATH_USERS)
    public Observable<Response<User>> createUser(@Body RequestWrapper user);

    @GET(PATH_USERS)
    public Observable<Response<List<User>>> getUsers(@Query("group_id") int groupId);

    @GET(PATH_USER)
    public Observable<Response<User>> getMe(@Path("id") String userId);

    @PATCH(PATH_USER)
    public Observable<Response<User>> updateUser(@Path("id") int userId, @Body UpdateRequestWrapper user);

    @PATCH(PATH_USER)
    public Observable<Response<User>> registerAccount(@Path("id") int userId, @Body UpdateAccountRequestWrapper user);

    @FormUrlEncoded
    @POST(PATH_LOGIN)
    public Observable<Response<User>> getAuthToken(@Field("account") String account, @Field("password") String password);

    @GET(PATH_USER_SEARCH)
    public Observable<Response<User>> search(@Query("account") String account);

    @PATCH(PATH_PASSWORDS)
    public Observable<Response<User>> changePassword(@Body ChangePasswordRequestWrapper user);

    @POST(PATH_PASSWORDS_INIT)
    public Observable<Response<ResponseMessage>> forgetPassword(@Body ForgetRequestWrapper user);

    @PATCH(PATH_PASSWORDS_RESET)
    public Observable<Response<User>> resetPassword(@Body ResetPasswordRequestWrapper user);

    @POST(PATH_NOTIFICATION_TOKENS)
    public Observable<Response<User>> registerToken(@Body NotificationRequestWrapper notification);

    @PATCH(PATH_NOTIFICATION_TOKENS)
    public Observable<Response<User>> updateToken(@Body NotificationRequestWrapper notification);

    @DELETE(PATH_NOTIFICATION_TOKENS)
    public Observable<Response<User>> deleteToken(@Body NotificationRequestWrapper notification);

    @POST(PATH_OAUTH)
    public Observable<Response<User>> signInWithOAuth(@Body OAuthRequestWrapper oauthRegistration);

    public class RequestWrapper {
        public PostData user;

        public RequestWrapper(@NonNull final String account,
                              @NonNull final String email,
                              @NonNull final String password) {
            this.user = new PostData(account, email, password);
        }

        class PostData {
            public String account;
            public String email;
            public String password;

            public PostData(@NonNull final String account,
                            @NonNull final String email,
                            @NonNull final String password) {
                this.account = account;
                this.email = email;
                this.password = password;
            }
        }
    }

    public class UpdateRequestWrapper {
        public PatchData user;

        public UpdateRequestWrapper(@NonNull final String username,
                                    @NonNull final String email) {
            this.user = new PatchData(username, email);
        }

        class PatchData {
            public String username;
            public String email;

            public PatchData(@NonNull final String username,
                             @NonNull final String email) {
                this.username = username;
                this.email = email;
            }
        }
    }

    public class UpdateAccountRequestWrapper {
        public PatchData user;

        public UpdateAccountRequestWrapper(@NonNull final String account) {
            this.user = new PatchData(account);
        }

        class PatchData {
            public String account;

            public PatchData(@NonNull final String account) {
                this.account = account;
            }
        }
    }

    public class ChangePasswordRequestWrapper {
        public PatchData user;

        public ChangePasswordRequestWrapper(@NonNull final String password,
                                            @NonNull final String newPassword,
                                            @NonNull final String newPasswordConfirmation) {
            this.user = new PatchData(password, newPassword, newPasswordConfirmation);
        }

        class PatchData {
            public String password;
            public String newPassword;
            public String newPasswordConfirmation;

            public PatchData(@NonNull final String password,
                             @NonNull final String newPassword,
                             @NonNull final String newPasswordConfirmation) {
                this.password = password;
                this.newPassword = newPassword;
                this.newPasswordConfirmation = newPasswordConfirmation;
            }
        }
    }

    public class ForgetRequestWrapper {
        public PostData user;

        public ForgetRequestWrapper(@NonNull final CharSequence email) {
            this.user = new PostData(email);
        }

        class PostData {
            public CharSequence email;

            public PostData(@NonNull final CharSequence email) {
                this.email = email;
            }
        }
    }

    public class ResetPasswordRequestWrapper {
        public PatchData user;

        public ResetPasswordRequestWrapper(@NonNull final CharSequence resetPasswordToken,
                                           @NonNull final CharSequence newPassword,
                                           @NonNull final CharSequence newPasswordConfirmation) {
            this.user = new PatchData(resetPasswordToken, newPassword, newPasswordConfirmation);
        }

        class PatchData {
            public CharSequence resetPasswordToken;
            public CharSequence newPassword;
            public CharSequence newPasswordConfirmation;

            public PatchData(@NonNull final CharSequence resetPasswordToken,
                             @NonNull final CharSequence newPassword,
                             @NonNull final CharSequence newPasswordConfirmation) {
                this.resetPasswordToken = resetPasswordToken;
                this.newPassword = newPassword;
                this.newPasswordConfirmation = newPasswordConfirmation;
            }
        }
    }

    public class NotificationRequestWrapper {
        public Data notificationToken;

        public NotificationRequestWrapper(@NonNull final String deviceToken,
                                          @Nullable final String authDeviceToken) {
            this.notificationToken = new Data(deviceToken, authDeviceToken);
        }

        class Data {
            public final String deviceType = "android";
            public String deviceToken;
            public String authDeviceToken;

            public Data(@NonNull final String deviceToken,
                        @Nullable final String authDeviceToken) {
                this.deviceToken = deviceToken;
                this.authDeviceToken = authDeviceToken;
            }
        }
    }

    public class OAuthRequestWrapper {
        public Data oauthRegistration;

        public OAuthRequestWrapper(@Nonnull final String thirdPartyId,
                                   @Nonnull final int oauthId,
                                   @Nonnull final String snsHashId) {
            this.oauthRegistration = new Data(thirdPartyId, oauthId, snsHashId);
        }

        class Data {
            public String thirdPartyId;
            public int oauthId;
            public String snsHashId;

            public Data(@Nonnull final String thirdPartyId,
                        @Nonnull final int oauthId,
                        @Nonnull final String snsHashId) {
                this.thirdPartyId = thirdPartyId;
                this.oauthId = oauthId;
                this.snsHashId = snsHashId;
            }
        }

    }
}

File: app/src/main/java/app/android/ttp/mikazuki/yoshinani/repository/retrofit/service/RetrofitUserService.kt
package app.android.ttp.mikazuki.yoshinani.repository.retrofit.service

import app.android.ttp.mikazuki.yoshinani.repository.retrofit.entity.ResponseMessage
import app.android.ttp.mikazuki.yoshinani.repository.retrofit.entity.User
import retrofit2.Response
import retrofit2.http.*
import rx.Observable

/**
 * @author haijimakazuki
 */
interface RetrofitUserService {

    @POST(PATH_USERS)
    fun createUser(@Body user: RequestWrapper): Observable<Response<User>>

    @GET(PATH_USERS)
    fun getUsers(@Query("group_id") groupId: Int): Observable<Response<List<User>>>

    @GET(PATH_USER)
    fun getMe(@Path("id") userId: String): Observable<Response<User>>

    @PATCH(PATH_USER)
    fun updateUser(@Path("id") userId: Int, @Body user: UpdateRequestWrapper): Observable<Response<User>>

    @PATCH(PATH_USER)
    fun registerAccount(@Path("id") userId: Int, @Body user: UpdateAccountRequestWrapper): Observable<Response<User>>

    @FormUrlEncoded
    @POST(PATH_LOGIN)
    fun getAuthToken(@Field("account") account: String, @Field("password") password: String): Observable<Response<User>>

    @GET(PATH_USER_SEARCH)
    fun search(@Query("account") account: String): Observable<Response<User>>

    @PATCH(PATH_PASSWORDS)
    fun changePassword(@Body user: ChangePasswordRequestWrapper): Observable<Response<User>>

    @POST(PATH_PASSWORDS_INIT)
    fun forgetPassword(@Body user: ForgetRequestWrapper): Observable<Response<ResponseMessage>>

    @PATCH(PATH_PASSWORDS_RESET)
    fun resetPassword(@Body user: ResetPasswordRequestWrapper): Observable<Response<User>>

    @POST(PATH_NOTIFICATION_TOKENS)
    fun registerToken(@Body notification: NotificationRequestWrapper): Observable<Response<User>>

    @PATCH(PATH_NOTIFICATION_TOKENS)
    fun updateToken(@Body notification: NotificationRequestWrapper): Observable<Response<User>>

    @DELETE(PATH_NOTIFICATION_TOKENS)
    fun deleteToken(@Body notification: NotificationRequestWrapper): Observable<Response<User>>

    @POST(PATH_OAUTH)
    fun signInWithOAuth(@Body oauthRegistration: OAuthRequestWrapper): Observable<Response<User>>

    class RequestWrapper(account: String,
                         email: String,
                         password: String) {
        internal var user: PostData

        init {
            this.user = PostData(account, email, password)
        }

        internal inner class PostData(var account: String,
                                      var email: String,
                                      var password: String)
    }

    class UpdateRequestWrapper(username: String,
                               email: String) {
        internal var user: PatchData

        init {
            this.user = PatchData(username, email)
        }

        internal inner class PatchData(var username: String,
                                       var email: String)
    }

    class UpdateAccountRequestWrapper(account: String) {
        internal var user: PatchData

        init {
            this.user = PatchData(account)
        }

        internal inner class PatchData(var account: String)
    }

    class ChangePasswordRequestWrapper(password: String,
                                       newPassword: String,
                                       newPasswordConfirmation: String) {
        internal var user: PatchData

        init {
            this.user = PatchData(password, newPassword, newPasswordConfirmation)
        }

        internal inner class PatchData(var password: String,
                                       var newPassword: String,
                                       var newPasswordConfirmation: String)
    }

    class ForgetRequestWrapper(email: CharSequence) {
        internal var user: PostData

        init {
            this.user = PostData(email)
        }

        internal inner class PostData(var email: CharSequence)
    }

    class ResetPasswordRequestWrapper(resetPasswordToken: CharSequence,
                                      newPassword: CharSequence,
                                      newPasswordConfirmation: CharSequence) {
        internal var user: PatchData

        init {
            this.user = PatchData(resetPasswordToken, newPassword, newPasswordConfirmation)
        }

        internal inner class PatchData(var resetPasswordToken: CharSequence,
                                       var newPassword: CharSequence,
                                       var newPasswordConfirmation: CharSequence)
    }

    class NotificationRequestWrapper(deviceToken: String,
                                     authDeviceToken: String?) {
        internal var notificationToken: Data

        init {
            this.notificationToken = Data(deviceToken, authDeviceToken)
        }

        internal inner class Data(var deviceToken: String,
                                  var authDeviceToken: String?) {
            val deviceType = "android"
        }
    }

    class OAuthRequestWrapper(thirdPartyId: String,
                              oauthId: Int,
                              snsHashId: String) {
        internal var oauthRegistration: Data

        init {
            this.oauthRegistration = Data(thirdPartyId, oauthId, snsHashId)
        }

        internal inner class Data(var thirdPartyId: String,
                                  var oauthId: Int,
                                  var snsHashId: String)

    }

    companion object {

        const val PATH_USERS = "users"
        const val PATH_USER = "users/{id}"
        const val PATH_LOGIN = "users/sign_in"
        const val PATH_USER_SEARCH = "users/search"
        const val PATH_PASSWORDS = "passwords"
        const val PATH_PASSWORDS_INIT = "passwords/init"
        const val PATH_PASSWORDS_RESET = "passwords/reset"
        const val PATH_NOTIFICATION_TOKENS = "notification_tokens"
        const val PATH_OAUTH = "oauth_registrations"
    }
}
